conans/client/tools/scm.py

```diff
@@ -89,23 +99,45 @@ def run(self, command):
         command = self._configure_ssl_verify + command
         return super(Git, self).run(command)
 
-    def clone(self, url, branch=None, args=""):
+    def clone(self, url, branch=None, args="", shallow=False):
+        """
+        :param url: repository remote URL to clone from (e.g. https, git or local)
+        :param branch: actually, can be any valid git ref expression like,
+        - None, use default branch, usually it's "master"
+        - branch name
+        - tag name
+        - revision sha256
+        - expression like HEAD~1
+        :param args: additional arguments to be passed to the git command (e.g. config args)
+        :param shallow:
+        :return: output of the clone command
+        """
+        # TODO: rename "branch" -> "element" in Conan 2.0
         url = self.get_url_with_credentials(url)
         if os.path.exists(url):
             url = url.replace("\\", "/")  # Windows local directory
-        if os.path.exists(self.folder) and os.listdir(self.folder):
+        if (os.path.exists(self.folder) and os.listdir(self.folder)) or (shallow and branch):
             if not branch:
                 raise ConanException("The destination folder '%s' is not empty, "
                                      "specify a branch to checkout (not a tag or commit) "
                                      "or specify a 'subfolder' "
                                      "attribute in the 'scm'" % self.folder)
             output = self.run("init")
             output += self.run('remote add origin "%s"' % url)
-            output += self.run("fetch ")
-            output += self.run("checkout -t origin/%s" % branch)
+            if shallow:
+                try:
+                    output += self.run('fetch --depth 1 origin "%s"' % branch)
+                    output += self.run('reset --hard FETCH_HEAD')
```

Comment:
⚠️  This will modify (and revert) any local changes. I can't remember all the use-cases that can lead us to this piece of code, so are we sure? Can it be called in the local workflow and drop all the local changes?

Reply:
I don't know what's the use-case of cloning into the existing directory, if you could add some backgorund, it would be helpful.
also, do you think it should use `--mixed` or `--soft` reset here?

Reply:
I think this can be executed during the `conan source ....` command, but not sure without running some tests... but if this command is being used on the local folder we cannot run a `git reset`.

Reply:
I am adding an additional test to validate the use-case of having local modifications (e.g. by `exports` or `exports_sources` in conanfile).
actually, the right option seems to be `reset --merge`:
```
--merge
Resets the index and updates the files in the working tree that are different between <commit> and HEAD, but keeps those which are different between the index and working tree (i.e. which have changes which have not been added). If a file that is different between <commit> and the index has unstaged changes, reset is aborted.
```
(from https://git-scm.com/docs/git-reset)
this way it works for having local file

Reply:
That `resets the index` is what scares me most: if it runs in my local copy and I have commits not pushed to the remotes, what will happen?

Reply:
okay, I've changed that to do just `git checkout FETCH_HEAD` instead of `git reset FETCH_HEAD`, so it's in align with regular (non-shallow) clone.